cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/ConsumerWorkingSetOperations.java

```diff
@@ -0,0 +1,193 @@
+    }
+    workingSet.setPartitionConsumerState(result.getPartitionConsumerState());
+  }
+
+  /**
+   * This method must be called on any partitions returned by the {@code #consumePartitions} method.
+   *
+   * @param partitionKeys list of partition keys to mark as either succeeded or failed processing
+   * @param succeeded whether or not processing of the specified partitions was successful
+   */
+  void onFinish(List<PartitionKey> partitionKeys, boolean succeeded) {
+    if (succeeded) {
+      commit(partitionKeys);
+    } else {
+      abort(partitionKeys);
+    }
+  }
+
+  // Removes the given partition keys from the working set, as they have been successfully processed.
+  private void commit(List<PartitionKey> partitionKeys) {
+    long expiryTime = getExpiryBorder();
+
+    for (PartitionKey key : partitionKeys) {
+      ConsumablePartition consumablePartition = workingSet.lookup(key);
+      assertInProgress(consumablePartition);
+      assertNotExpired(consumablePartition, expiryTime);
+      workingSet.remove(key);
+    }
+  }
+
+  // Resets the process state of the given partition keys, as they were not successfully processed.
+  private void abort(List<PartitionKey> partitionKeys) {
+    long expiryTime = getExpiryBorder();
+
+    List<PartitionKey> discardedPartitions = new ArrayList<>();
+    for (PartitionKey key : partitionKeys) {
+      ConsumablePartition consumablePartition = workingSet.lookup(key);
+      assertInProgress(consumablePartition);
+      assertNotExpired(consumablePartition, expiryTime);
+      // either reset its processState, or remove it from the workingSet, depending on how many tries it already has
+      if (consumablePartition.incrementNumFailures() < consumerConfiguration.getNumRetries()) {
+        consumablePartition.resetState();
+      } else {
+        discardedPartitions.add(key);
+        workingSet.remove(key);
+      }
+    }
+    if (!discardedPartitions.isEmpty()) {
+      LOG.warn("Discarded keys due to being retried {} times: {}",
+               consumerConfiguration.getNumRetries(), discardedPartitions);
+    }
+  }
+
+  // ensure that user doesn't try to commit/abort a partition that isn't in progress
+  private void assertInProgress(ConsumablePartition consumablePartition) {
+    if (!consumablePartition.isInProgress()) {
+      throw new IllegalStateException(String.format("Partition not in progress: %s",
+                                                    consumablePartition.getPartitionKey()));
+    }
+  }
+
+  // ensure that user doesn't try to commit/abort a partition that is expired
+  private void assertNotExpired(ConsumablePartition consumablePartition, long expiryBorder) {
+    if (consumablePartition.getTimestamp() < expiryBorder) {
+      throw new IllegalStateException(String.format("Attempting to operate on an expired partition: %s",
+                                                    consumablePartition.getPartitionKey()));
+    }
```

Comment:
So when does the "not-in-progress" vs the "expired" case happen? If I understand correctly: "not-in-progress" happens if the doExpiry() has marked it as expired and that got persisted. "expired" case happens if the timeout is only detected now. But both can represent a timeout, right? I think we need clear semantics for what exceptions are thrown under what circumstances. It seems here it is by chance, whether a concurrent consumer has already marked the partition as expired or not, we throw a different exception.